### PR TITLE
Leverage PHP 7 language features

### LINKED ISSUE #7597: Leverage PHP 7 language features
## Task Description

With the PHP version requirement bump we can use language features introduced in PHP 7, like scalar type declarations, return type declarations, and the null coalescing operator.

See https://www.php.net/manual/en/migration70.new-features.php for a full list of changes.

I expect the IDE can help with a lot of these optimizations.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Use the library [rector](https://github.com/rectorphp/rector ) to upgrade to PHP 7 language features. Features implemented. 

- Return type hints.
- Nullable operators. 
- Grouped use statements. 

Replacement for https://github.com/google/web-stories-wp/pull/8264. 

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7597

### CODE REVIEW

**Comment by swissspidy** [comment]
A bit difficult now with discussion now split between two PRs, but regarding your [last comment](https://github.com/google/web-stories-wp/pull/8264#discussion_r668916013):

> seems like the issue is really param hints

In this PR you only added return type hints and tests are still failing. So either seems to be an issue. Which means we have to solve this some other way 😞 

Regarding your [second to last comment](https://github.com/google/web-stories-wp/pull/8264#discussion_r668750072):

> Are you recommend that do option 1.

I was not recommending a specific option, but rather suggesting that we should explore other options like the one shared there.

I also brought up a third option to just patch PHPUnit 6 instead, but no idea how feasible that would be (would need to be explored)

> If so, I think that has some bigger issues. If we commit PHPUnit 7.x and it's dependencies to lock file, it would mean, everywhere we need to code to run in PHP 7.0, like building the plugin, we would need to downgrade PHPUnit again.

I don't think we'd need to downgrade PHPUnit everywhere else, but the problem is probably that we need to use `--ignore-platform-reqs` everywhere. So yeah, not really feasible.

> Another idea would be maintain two composer files.

I don't think composer allows switching files like that, but a different `composer.json` in a different directory is doable (using the `-d`/`--working-dir` flag).

See #4823 for some prior art.

This _might_ work. but not sure about implications for maintainers.

----

**tl;dr** Overall none of the options seem _super_ compelling and some further exploration / proof-of-concepts for each option would be helpful to make a decision.

File: patches/phpunit-mock-objects.patch [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+===================================================================
+diff --git a/src/Generator.php b/src/Generator.php
+--- a/src/Generator.php	(revision 0dfddc236629eb7ead4df2dfa3d79d548e011a3b)
++++ b/src/Generator.php	(date 1626259541863)
+@@ -959,7 +959,7 @@
+         }
+ 
+         if ($method->hasReturnType()) {
+-            $returnType = (string) $method->getReturnType();
++            $returnType = PHP_VERSION_ID >= 70100 ? $method->getReturnType()->getName() : (string) $method->getReturnType();
```
Comment (swissspidy) [inline patches/phpunit-mock-objects.patch:14]: Amazing!!!! 🎉 🥳

**Comment by spacedmonkey** [comment]
@swissspidy Can you review again? 

I patch the phpunit error in 7.4 and it works a treat.

**Comment by swissspidy** [comment]
Not all functions seem to have param type declarations, for example `rest_preload_api_request()` doesn't. What's the reason there?

**Comment by spacedmonkey** [comment]
> Not all functions seem to have param type declarations, for example rest_preload_api_request() doesn't. What's the reason there?

You mentioned to be, that not all functions / methods should have type hints. Any method that is hooked into a filter or hook into WordPress is a little of control, so I remove the type hints. 

I use the rector tools to find all the places it could. I went through as many files as I could, adding hints where the tool fails and removed others where it did not make sense. 

This is a just a start, as we touch and change the code, we can add / remove hints. This is just the first automated push for param/return hints.

**Comment by swissspidy** [comment]
Great. Just wanted to confirm!

**Review by swissspidy — APPROVED** [review]
Let's 🚢  it!
